+++
title = "16S流程重构建"
date = 2018-07-27T09:39:35+08:00
draft = false

# Tags: can be used for filtering projects.
# Example: `tags = ["machine-learning", "deep-learning"]`
tags = ['16S rDNA测序']

# Project summary to display on homepage.
summary = ""

# Optional image to display on homepage.
image_preview = ""

# Optional external URL for project (replaces project detail page).
external_link = ""

# Does the project detail page use math formatting?
math = false

# Does the project detail page use source code highlighting?
highlight = true

# Featured image
# Place your image in the `static/img/` folder and reference its filename below, e.g. `image = "example.jpg"`.
[header]
image = ""
caption = ""

+++

## 第一部分： reads -> contigs并统计

脚本start2SummaryContigs.batch的内容：

```shell
make.file(inputdir=., type=fastq, prefix=stability)
make.contigs(file=stability.files, processors=56)
summary.seqs(fasta=stability.trim.contigs.fasta)
```

使用时，可以比如`nohup mothur start2SummaryContigs.batch > step1.log &`这样。

之所以选择在这里截断流程一次，是因为summary之后需要根据统计得到的reads的长度分布去修改下一阶段使用的脚本内的参数（比如`maxlength`）。

## 第二部分： 提取高变区对应的参考序列

这部分参考[上次组会汇报的内容]({{< ref "talk/20180721-16s.zh.md" >}})。

## 第三部分： 过滤并统计

脚本screen2SummaryCounts.batch的内容：

```shell
screen.seqs(fasta=stability.trim.contigs.fasta, group=stability.contigs.groups, summary=stability.trim.contigs.summary, maxambig=0, maxlength=475)
summary.seqs()
unique.seqs(fasta=stability.trim.contigs.good.fasta)
count.seqs(name=stability.trim.contigs.good.names, group=stability.contigs.good.groups)
summary.seqs(count=stability.trim.contigs.good.count_table)
```



这里我有一个失误。

完成align之后，是想要summary之后对比对结果进行过滤的：

```shell
mothur > summary.seqs(fasta=stability.trim.contigs.good.unique.align, count=stability.trim.contigs.good.count_table)
```

报错了：
```pre
[ERROR]: Your count file contains 6400841 unique sequences, but your fasta file contains 3777293. File mismatch detected, quitting command.
```

分别对前面的结果做了一下summary，原来unique之后和比对的结果的序列数差了差不多一半。回去查一下`align.seqs()`的报告，原来很多都没有比对上（原因待查，V45的区域应该没拿错。至于为什么暂时我也不清楚），这样的话，考虑把count table剪一剪拿去用。

看下count table的格式：

```pre
Representative_Sequence total   LZW113
FCAVD1Y_1_1112_19272_5098       1       1
FCAVD1Y_1_2105_6942_8265        1037    1037
FCAVD1Y_1_1113_27631_16492      1       1
FCAVD1Y_1_2110_13690_19012      1       1
FCAVD1Y_1_2113_9095_17142       12      12
FCAVD1Y_1_2107_11955_8783       1       1
FCAVD1Y_1_1104_12000_5514       1       1
FCAVD1Y_1_1113_19468_4747       1       1
FCAVD1Y_1_2102_16601_17755      1       1
```

也看下alignment的ID格式（类fasta文件）：

```shell
grep ">" stability.trim.contigs.good.unique.align |head
```

```pre
>FCAVD1Y_1_1112_19272_5098
>FCAVD1Y_1_2105_6942_8265
>FCAVD1Y_1_1113_27631_16492
>FCAVD1Y_1_2110_13690_19012
>FCAVD1Y_1_2113_9095_17142
>FCAVD1Y_1_2107_11955_8783
>FCAVD1Y_1_1104_12000_5514
>FCAVD1Y_1_1113_19468_4747
>FCAVD1Y_1_2102_16601_17755
>FCAVD1Y_1_2103_6545_6854
```

这样就好办了。截取那个table，只保留ID共有的部分:

```shell
perl -lanE 'if ($switch) {say if (/^Re/ || exists $need{$F[0]})} else {$need{$1}++ if />(.+)$/} $switch++ if eof(ARGV)' stability.trim.contigs.good.unique.align stability.trim.contigs.good.count_table > stability.trim.contigs.good.filter.count_table
mv stability.trim.contigs.good.count_table stability.trim.contigs.good.count_table.bak
mv stability.trim.contigs.good.filter.count_table stability.trim.contigs.good.count_table
```

然后再去mothur里面搞：

```shell
mothur > summary.seqs(fasta=stability.trim.contigs.good.unique.align, count=stability.trim.contigs.good.count_table)
```

```pre
                Start   End     NBases  Ambigs  Polymer NumSeqs
Minimum:        0       0       0       0       1       1
2.5%-tile:      2       11344   299     0       4       124826
25%-tile:       2       11344   302     0       4       1248258
Median:         2       11344   304     0       4       2496516
75%-tile:       2       11344   304     0       5       3744774
97.5%-tile:     2       11344   307     0       7       4868206
Maximum:        11344   11344   405     0       9       4993031
Mean:   47      11322   301     0       4
# of unique seqs:       3777293
total # of seqs:        4993031
```

```shell
screen.seqs(fasta=stability.trim.contigs.good.unique.align, count=stability.trim.contigs.good.count_table, summary=stability.trim.contigs.good.unique.summary, start=2, end=11344, maxhomop=8)
```

> 这个screen的过程，其实在这个数据里面是没必要指定start和end了。但是为了之后的流程，我先标出来。